File: bidragsevne/src/main/java/no/nav/bidrag/beregn/felles/bidragsevne/periode/Periodiserer.java
package no.nav.bidrag.beregn.felles.bidragsevne.periode;

import static java.util.stream.Collectors.toList;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Set;
import no.nav.bidrag.beregn.felles.bidragsevne.periode.grunnlag.Periode;
import no.nav.bidrag.beregn.felles.bidragsevne.periode.grunnlag.PeriodisertGrunnlag;

public class Periodiserer {

  private final Set<LocalDate> bruddpunkter = new HashSet<>();
  private boolean aapenSluttdato = false;

  public Periodiserer addBruddpunkt(LocalDate dato) {
    bruddpunkter.add(dato);
    return this;
  }

  private Periodiserer addBruddpunkter(Periode periode) {
    addBruddpunkt(periode.getDatoFra());
    if (periode.getDatoTil() == null) {
      aapenSluttdato = true;
    } else {
      addBruddpunkt(periode.getDatoTil());
    }
    return this;
  }

  public Periodiserer addBruddpunkter(PeriodisertGrunnlag grunnlag) {
    addBruddpunkter(grunnlag.getDatoFraTil());
    return this;
  }

  Periodiserer addBruddpunkter(Iterable<? extends PeriodisertGrunnlag> grunnlagListe) {
    for (PeriodisertGrunnlag grunnlag : grunnlagListe) {
      addBruddpunkter(grunnlag);
    }
    return this;
  }

  // Genererer brudd når søknadsbarnet passerer 11 år og 18 år
  // 0 og 11 år justeres til den første inneværende måned, 18 år justeres til den første neste måned
//  Periodiserer addBruddpunkter(LocalDate fodselDato, LocalDate beregnDatoFra, LocalDate beregnDatoTil) {
//
//    var barn11AarDato = fodselDato.plusYears(11).with(firstDayOfMonth());
//    var barn18AarDato = fodselDato.plusYears(18).with(firstDayOfNextMonth());
//    var barn11AarIPerioden = barn11AarDato.isAfter(beregnDatoFra.minusDays(1)) && barn11AarDato.isBefore(beregnDatoTil.plusDays(1));
//    var barn18AarIPerioden = barn18AarDato.isAfter(beregnDatoFra.minusDays(1)) && barn18AarDato.isBefore(beregnDatoTil.plusDays(1));
//
//    if (barn11AarIPerioden) {
//      addBruddpunkt(barn11AarDato);
//    }
//
//    if (barn18AarIPerioden) {
//      addBruddpunkt(barn18AarDato);
//    }
//    return this;
//  }

  public List<Periode> finnPerioder(LocalDate beregnDatoFra, LocalDate beregnDatoTil) {
    var sortertBruddpunktListe = bruddpunkter.stream().filter((dato) -> dato.isAfter(beregnDatoFra.minusDays(1)))
        .filter((dato) -> dato.isBefore(beregnDatoTil.plusDays(1))).sorted().collect(toList());

    List<Periode> perioder = new ArrayList<>();
    Iterator<LocalDate> bruddpunktIt = sortertBruddpunktListe.iterator();
    if (bruddpunktIt.hasNext()) {
      LocalDate start = bruddpunktIt.next();
      while (bruddpunktIt.hasNext()) {
        LocalDate end = bruddpunktIt.next();
        perioder.add(new Periode(start, end));
        start = end;
      }
      if (aapenSluttdato) {
        perioder.add(new Periode(start, null));
      }
    }
    return perioder;
  }


}

File: src/main/kotlin/no/nav/bidrag/beregn/felles/periode/Periodiserer.kt
package no.nav.bidrag.beregn.felles.periode

import no.nav.bidrag.beregn.felles.bo.Periode
import no.nav.bidrag.beregn.felles.bo.PeriodisertGrunnlag
import java.time.LocalDate
import java.time.temporal.TemporalAdjusters
import java.util.ArrayList
import java.util.HashSet
import java.util.stream.Collectors

class Periodiserer {
    internal val bruddpunkter: MutableSet<LocalDate> = HashSet()
    internal var aapenSluttdato = false

    fun addBruddpunkt(dato: LocalDate): Periodiserer {
        bruddpunkter.add(dato)
        return this
    }

    private fun addBruddpunkter(periode: Periode) {
        addBruddpunkt(periode.datoFra)

        if (periode.datoTil == null) {
            aapenSluttdato = true
        } else {
            addBruddpunkt(periode.datoTil)
        }
    }

    fun addBruddpunkter(grunnlag: PeriodisertGrunnlag): Periodiserer {
        addBruddpunkter(grunnlag.getDatoFraTil())
        return this
    }

    fun addBruddpunkter(grunnlagListe: Iterable<PeriodisertGrunnlag>): Periodiserer {
        for (grunnlag in grunnlagListe) {
            addBruddpunkter(grunnlag)
        }

        return this
    }

    // Genererer brudd når søknadsbarnet passerer 11 år og 18 år
    // 0 og 11 år justeres til den første inneværende måned, 18 år justeres til den første neste måned
    fun addBruddpunkter(fodselDato: LocalDate, beregnDatoFra: LocalDate, beregnDatoTil: LocalDate): Periodiserer {
        val barn11AarDato = fodselDato.plusYears(11).with(TemporalAdjusters.firstDayOfMonth())
        val barn18AarDato = fodselDato.plusYears(18).with(TemporalAdjusters.firstDayOfNextMonth())
        val barn11AarIPerioden = barn11AarDato.isAfter(beregnDatoFra.minusDays(1)) && barn11AarDato.isBefore(beregnDatoTil.plusDays(1))
        val barn18AarIPerioden = barn18AarDato.isAfter(beregnDatoFra.minusDays(1)) && barn18AarDato.isBefore(beregnDatoTil.plusDays(1))

        if (barn11AarIPerioden) {
            addBruddpunkt(barn11AarDato)
        }

        if (barn18AarIPerioden) {
            addBruddpunkt(barn18AarDato)
        }

        return this
    }

    // Setter perioder basert på fra- og til-dato
    fun finnPerioder(beregnDatoFra: LocalDate, beregnDatoTil: LocalDate): List<Periode> {
        val sortertBruddpunktListe = bruddpunkter.stream()
                .filter { dato: LocalDate -> dato.isAfter(beregnDatoFra.minusDays(1)) }
                .filter { dato: LocalDate -> dato.isBefore(beregnDatoTil.plusDays(1)) }
                .sorted().collect(Collectors.toList())

        val perioder: MutableList<Periode> = ArrayList()
        val bruddpunktIt = sortertBruddpunktListe.iterator()

        if (bruddpunktIt.hasNext()) {
            var start: LocalDate? = bruddpunktIt.next()

            while (bruddpunktIt.hasNext()) {
                val end = bruddpunktIt.next()
                perioder.add(Periode(start!!, end))
                start = end
            }

            if (aapenSluttdato) {
                perioder.add(Periode(start!!, null))
            }
        }

        return perioder
    }
}
